ocaml/typing/signature_with_global_bindings.ml

```diff
@@ -0,0 +1,74 @@
+[@@@ocaml.warning "+a-9-40-41-42"]
+
+type t = {
+  sign : Subst.Lazy.signature;
+  bound_globals : Global_module.t array;
+}
+
+let read_from_cmi (cmi : Cmi_format.cmi_infos_lazy) =
+  let sign =
+    (* Freshen identifiers bound by signature *)
+    Subst.Lazy.signature Make_local Subst.identity cmi.cmi_sign in
+  let bound_globals = cmi.cmi_globals in
+  { sign; bound_globals }
+
+let array_fold_left_filter_map f init array =
+  let ans, new_array = Array.fold_left_map f init array in
+  let new_array =
+    (* To be replaced with something faster if we need it *)
+    Array.of_seq (Seq.filter_map (fun a -> a) (Array.to_seq new_array))
+  in
+  ans, new_array
+
+let subst t (args : (Global_module.Name.t * Global_module.t) list) =
+  let { sign; bound_globals } = t in
+  match args with
+  | [] -> t
+  | _ ->
+      (* The global-level substitution *)
+      let arg_subst = Global_module.Name.Map.of_list args in
+      (* Take a bound global, substitute arguments into it, then return the
+         updated global while also adding it to the term-level substitution *)
+      let add_and_update_binding subst bound_global =
+        let name = Global_module.to_name bound_global in
+        if Global_module.Name.Map.mem name arg_subst then
+          (* This will be handled by [add_arg] below *)
+          subst, None
+        else
+          begin
+            let value, changed = Global_module.subst bound_global arg_subst in
+            let name_id = Ident.create_global name in
+            let value_as_name = Global_module.to_name value in
+            let value_id = Ident.create_global value_as_name in
```

Comment:
I don't think it's avoidable here: we genuinely need the `Global_module.t` so we can call `Global_module.is_complete` and we genuinely need the `Global_module.Name.t` so we can put it back into an `Ident.t`.

Reply:
As discussed offline, we think it might be worthwhile to explore the avenue where `Global_module.Name.t` is similar to `Longident.t` and `Global_module.t` is part of `Path.t`. The intuition here is that the former is user syntax while the latter is typed (and explicitly contains everything, including hidden args). That would allow `Global_module.t` to be recognized and processed by `subst.ml`, and we wouldn't need this file.